### ISSUE TITLE
Solving a complex-valued version of Time dependent Heat Eq. (Example 16)

### ISSUE THREAD

**Issue by jchan192** [mfem/mfem#2674]
Hi all,

I am trying to modify example ex16.cpp into a complex value version, which can be used to solve the linear Schrodinger Equation,
but I am facing trouble while using the `SesquilinearForm` classes. 

In ex16.cpp,  `M->FormSystemMatrix(ess_tdof_list, Mmat);`  cannot be directly applied to the case with complex valued
because `M = new SesquilinearForm(&fespace);` does not have the member function `FormSystemMatrix (const Array< int > &ess_tdof_list, OpType &A)`, but only `FormSystemMatrix (const Array< int > &ess_tdof_list, OperatorHandle &A)`.

Is there a work around to this problem? I have checked ex22.cpp and ex25.cpp, but I dont quite understand how it could help me to resolve this problem. 

Thank you for your time

**Comment by v-dobrev** [comment]
You should be able to change the type of `Mmat` from `HypreParMatrix` to `OperatorHandle`. After calling `ParSesquilinearForm::FormSystemMatrix(const Array<int> &ess_tdof_list, OperatorHandle &A)`, the `OperatorHandle` will store a pointer to a `ComplexHypreParMatrix` (unless you call `ParSesquilinearForm::SetAssemblyLevel` with `assembly_level != AssemblyLevel::LEGACY`). To access this `ComplexHypreParMatrix` from the `OperatorHandle Mmat`, you can use:
```c++
   Mmat.As<ComplexHypreParMatrix>()->real();   // this is a HypreParMatrix &
```
or you can create a temporary pointer and use that:
```c++
   ComplexHypreParMatrix *M_ptr;
   Mmat.Get(M_ptr);
   M_ptr->real();   // this is a HypreParMatrix &

```

**Comment by jchan192** [comment]
Thanks for the reply.
Sorry that If I didnt explain in detail, but the `Mmat`  in ex16.cpp is declared as `SparseMatrix Mmat;` instead of `OperatorHandle`.
In this case, the answer will be a bit different then?

**Comment by v-dobrev** [comment]
Yes, in serial, `SesquilinearForm::FormSystemMatrix` will return an `OperatorHandle` that stores a `ComplexSparseMatrix` (instead of `ComplexHypreParMatrix`). The `real()` and `imag()` methods of that class will return `SparseMatrix &` (instead of `HypreParMatrix &`).

**Comment by jchan192** [comment]
@v-dobrev 
Thanks again.  I follow approximately what you mentioned and implement the following
```
   M = new SesquilinearForm(&fespace);                                                                                                                                                                             
   M->AddDomainIntegrator(new MassIntegrator(),new MassIntegrator());                                                                                                                                              
   M->Assemble();                                                                                                                                                                                                  
   M->FormSystemMatrix(ess_tdof_list, Mmatc);                                                                                                                                                                      
   Mmat = Mmatc.As<ComplexSparseMatrix>()->imag();      

```
 it compiles normally. But the result is not evolving as what I expected.

The real part of the of the `ComeplexGridFunction u` evolves as expected, but the imaginary part  still does not evolve correctly. 
I think the problem is in the `ODEsolver ode_solver(u, t, dt)` and `Mult(u, dudt)`, which only accepts u as vector.  

I could try to convert the vector into complex vector, but I wonder if you could provide me some advise regarding how to modify the code correctly.

**Comment by jchan192** [comment]
So I started implementing a new complex ode solver in `ode.cpp`. 
I introduced a new member `CStep` to the Class `ODESolver`
Here is a simple code of the Complex version of Euler forward method.

```
void SP_ForwardEulerSolver::CStep(Vector &x_R, Vector &x_I, double &t, double &dt)                                                                                                                                 
{                                                                                                                                                                                                                  
                                                                                                                                                                                                              
  std::cout << "I am here!\n";                                                                                                                                                                                     
  f->SetTime(t);                                                                                                                                                                                                   
                                                                                                                                                                                                                   
  f->Mult(x_R, dxdt_R);                                                                                                                                                                                            
  f->Mult(x_I, dxdt_I);                                                                                                                                                                                            
                                                                                                                                                                                                                   
  x_I.Add(dt, dxdt_I);         // have segmentation fault when comment this , or next line                                                                                                                                                                                    
  x_R.Add(dt, dxdt_R);                                                                                                                                                                                             
                                                                                                                                                                                                                   
  t += dt;                                                                                                                                                                                                                                                                                                                                                                                                                
}       
```
I found that the code runs

**Comment by v-dobrev** [comment]
You should be able to use `ComplexSparseMatrix` directly with the ODE solvers. This operator uses complex vector represented as
```ascii
x = / x_real \
    \ x_imag /
```
See the general description of `ComplexSparseMatrix`'s base class, `ComplexOperator`: https://github.com/mfem/mfem/blob/539f663fe6ad3738b6de9615f6c74dfa44927578/linalg/complex_operator.hpp#L28-L68

**Comment by jchan192** [comment]
Thanks for the reply. 
Somehow I managed to make it work by seperating the complex function into real and imaginary parts. 
I will concluded that it is indeed possible to modify ex16 into complex value problems.

**Comment by IhCosmo** [comment]
Hi there,

I am also interested in solving the Schrodinger equation using MFEM. I have tried to follow the discussion with `
 ComplexSparseMatrix *Mmat
 ` and
```
   M = new SesquilinearForm(&fespace);
   M->AddDomainIntegrator(new MassIntegrator(),new MassIntegrator());
   M->Assemble();
   M->FormSystemMatrix(ess_tdof_list, Mmat_op);
   Mmat = Mmat_op.As<ComplexSparseMatrix>();
```

However, the first problem is with the preconditioner, 

```
M_solver.SetPreconditioner(M_prec);
SparseSmoother::SetOperator : not a SparseMatrix!
```


After commenting this line out the code compiles. Assuming that we solved (somehow) the previous problem, the Implicit Solve requires using the operator `Add`. However, I don't understand how to work this out. I thought maybe I should work with `ComplexSparseMatrix *T; // T = M + dt K` and try to use `Add`  in every component `T->real() = Add(...)`, but it is not working.

Should I continue like this?


And finally, from a previous answer, I didn't understand how to use the `ComplexSparseMatrix` to create a kind of ComplexVector. Is it something like the `ComplexGridFunction`?


Sorry for a lot of questions; and thanks for your time.
